### PR TITLE
fix(unplugin): serve the placeholder stylesheet under Vite's bundled dev server

### LINKED ISSUE #1301: [BUG] unplugin: placeholder CSS never loads under Vite's bundled dev mode
### Describe the bug

With `useCssPlaceholder` and Vite's `experimental.bundledDev`, the stylesheet that carries the marker never receives the StyleX rules. JavaScript compiles and elements get their `sx-*` class names, but the page has no StyleX CSS, and edits do not update it either.

Bundled dev hands the module graph to Rolldown and compiles every imported stylesheet into JavaScript. Two things the plugin relies on are gone in that mode:

- There is no CSS asset in the output, so `generateBundle` finds nothing to inject into and logs the warning below.
- Vite skips `handleHotUpdate`, and the client module graph is not populated, so the refresh the plugin sends after late transforms (`moduleGraph.invalidateModule` plus a `css-update`) has nothing to act on.

Background: https://github.com/vitejs/vite/discussions/22226

### Error stack

```
StyleX: no CSS asset contained the placeholder marker, so no styles were injected.
```

### OS

macOS

### Browser

Chrome

### Compiler version

@stylexswc/rs-compiler 0.19.0-rc.1

### Bundler plugin(s) version

@stylexswc/unplugin 0.18.5 and 0.19.0-rc.1, vite 8.2.2

### To Reproduce

1. `vite.config.ts` with `experimental: { bundledDev: true }` and `stylexSwc({ useCssPlaceholder: '/* @stylex-placeholder */', rsOptions: { dev: true } })`.
2. `style.css` containing the marker, imported from a module that also uses `stylex.create`.
3. `vite dev`, open the page.

### Expected behavior

The stylesheet loads with the StyleX rules at the marker on first paint. Editing a StyleX module, a `defineVars` or `createTheme` module, or the stylesheet itself updates the page over HMR without a reload. Regular dev and production builds are unaffected.

### Additional context

A related gap in regular dev: when the marker stylesheet is referenced by a `<link>` in `index.html` rather than imported from JavaScript, the late-rules refresh never reaches it. Vite registers that stylesheet as a `?direct` module, which the plugin's collector and load hook both skip.

### PR DESCRIPTION
**Pull request by skovhus**
## Description

Under Vite's `experimental.bundledDev`, Rolldown compiles every imported stylesheet into JavaScript, so the placeholder stylesheet never gets the StyleX rules and nothing refreshes it. The stylesheet that carries the marker is now taken out of the bundle: its import resolves to a small runtime module that links it from the dev server, a middleware renders it from the file and the current rules, and one hot event refetches it when rules or the stylesheet change. Only the bundled client environment is redirected; other environments, SSR included, keep the regular path.

Two regular-dev fixes ride along: a marker stylesheet referenced only by a `<link>` in the HTML is now matched (it is registered as a `?direct` module), and the late-rules refresh sends both update kinds so linked and imported stylesheets are both reached.

Verified in a headless Chromium: first paint, style edits, `defineVars` and `createTheme` edits, modules added after the initial build, lazily bundled modules, stylesheet and `@import` partial edits, a non-root base, an out-of-root stylesheet, a server restart, regular dev, and a production build whose CSS is byte-identical to the unpatched plugin.

## Fixes

Fixes #1301

## Type of change

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [x] This change requires a documentation update

## Checklist

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] New and existing unit tests pass locally with my changes
- [x] I have read the [CONTRIBUTING](CONTRIBUTING.md) document

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

File: packages/unplugin/README.md [hunk 1]
```diff
@@ -360,8 +360,35 @@ The plugin replaces the marker with the generated StyleX CSS during the build.
 > So the rules are minified along with everything else, but per-module steps
 > such as PostCSS or Lightning CSS transpilation never run over them. Use
 > [`transformCss`](#transformcss) to run your own processing over the rules.
-> The dev server has no bundle step and inlines the rules in the stylesheet, so
-> there the pipeline sees everything.
+> In development the rules are put into the stylesheet before the pipeline
```
Comment (skovhus) [inline packages/unplugin/README.md:363]: Not sure if we want this here.

File: packages/unplugin/src/index.ts [hunk 2]
```diff
@@ -695,16 +732,39 @@ export const unpluginFactory: UnpluginFactory<UnpluginStylexRSOptions | undefine
...
+        const runtime = bundledDevCss.load(id);
+        if (runtime) return runtime;
         // Only handle CSS files with useCssPlaceholder
         if (!normalizedOptions.useCssPlaceholder) return null;
-        if (!id.endsWith('.css')) return null;
+        // A stylesheet the browser fetched through `<link>` arrives with the
+        // `?direct` query the dev server adds; every other query (`?inline`,
+        // `?url`, `?raw`) asks for something other than the stylesheet.
+        const [cssFile = id, query] = id.split('?');
+        if (!cssFile.endsWith('.css') || (query !== undefined && query !== 'direct')) {
```
Comment (Dwlad90) [inline packages/unplugin/src/index.ts]: **`load()`: a `<link>` href that has its own query does not get the marker replaced**

The new guard accepts a query only when the query is exactly `direct`:

```ts
const [cssFile = id, query] = id.split('?');
if (!cssFile.endsWith('.css') || (query !== undefined && query !== 'direct')) {
  return null;
}
```

Vite's `injectQuery` puts `direct` in front of the query that the href already has. So `<link href="/style.css?v=2">` comes to this hook as `style.css?direct&v=2`. Then `query` is `'direct&v=2'`, the guard returns `null`, and Vite serves the raw file. The marker (`@stylex;` or the placeholder comment) goes to the browser as text. No error is reported.

The collector change in this PR makes the effect worse. `invalidateAndCollectCssModules` now matches this module by `mod.file`. Each late-rules refresh sends a `css-update` for a stylesheet that never gets the rules. The browser fetches the same unfilled stylesheet again and again.

I reproduced the problem on this branch with `server.transformRequest` in regular dev:

```text
/stylex.css?direct       => marker left: false | has rules: true
/stylex.css?direct&v=1   => marker left: true  | has rules: false
/stylex.css?v=1&direct   => marker left: true  | has rules: false
```

This is not a regression against `develop`. The old `id.endsWith('.css')` test also skipped these ids. But the PR says: "a marker stylesheet referenced only by a `<link>` in the HTML is now matched". The *Additional context* section of #1301 asks for this fix. At the moment, the fix works only for an href without a query. A cache-busting query on a `<link>` is common. Users will hit this case.

**Suggested fix.** Use the same test as Vite's `isDirectRequest`. Do not compare the full query string.

```ts
// Vite's own `isDirectRequest`: the dev server puts `direct` in front of the
// query that the `<link>` href already has.
const DIRECT_REQUEST_RE = /(?:\?|&)direct(?:&|$)/;
```

```diff
-        const [cssFile = id, query] = id.split('?');
-        if (!cssFile.endsWith('.css') || (query !== undefined && query !== 'direct')) {
+        const queryStart = id.indexOf('?');
+        const cssFile = queryStart === -1 ? id : id.slice(0, queryStart);
+        if (!cssFile.endsWith('.css') || (queryStart !== -1 && !DIRECT_REQUEST_RE.test(id))) {
           return null;
         }
```

The change keeps the current behavior for the other cases. An id without a query and an id with `?direct` give the same result as before. `?inline`, `?url` and `?raw` still return `null`. I checked this: `?inline` still keeps the marker, and `?url` is not changed. The only new case that the guard accepts is `direct` together with another query.

Please add a test case for `/stylex.css?direct&v=1` to `vite.test.ts`. Then the collector and the load hook stay in agreement.

File: packages/unplugin/src/utils/bundledDevCss.ts [hunk 3]
```diff
@@ -0,0 +1,317 @@
...
+  ): void {
+    // The base is still on the URL here: this runs ahead of Vite's own base
+    // middleware, and the hrefs were built with the base for that reason.
+    const url = new URL(req.url ?? '/', 'http://localhost');
+    // A parent router in middleware mode strips its mount path before this
+    // runs, so an href is also matched by its tail.
+    const file =
+      fileByPathname.get(url.pathname) ??
+      [...fileByPathname].find(([href]) => href.endsWith(url.pathname))?.[1];
+    if (!file || !config || !marker || !url.searchParams.has(STYLESHEET_QUERY)) {
```
Comment (Dwlad90) [inline packages/unplugin/src/utils/bundledDevCss.ts]: **O1. This middleware parses the URL and scans the href map on every request.**

This handler runs before Vite's own middlewares. So it sees every module request, HMR request and asset request of the dev server. For each of them, it makes a `URL`, does a Map lookup, and then runs an `O(n)` `find` with a string test on each href. Only after that does it test for the `stylex-css` query, and most requests do not have that query.

`n` is small. But this is the hot path of the dev server. Do the cheap test first, and do the parse and the lookup only for the few requests that have the query. The `config` and `marker` tests can move into the same first gate.

```diff
-    const url = new URL(req.url ?? '/', 'http://localhost');
-    const file =
-      fileByPathname.get(url.pathname) ??
-      [...fileByPathname].find(([href]) => href.endsWith(url.pathname))?.[1];
-    if (!file || !config || !marker || !url.searchParams.has(STYLESHEET_QUERY)) {
+    // This runs for every request the dev server gets. The plain substring
+    // test comes first. The URL parse and the lookup run only for the few
+    // requests that name the query.
+    if (!config || !marker || !req.url?.includes(STYLESHEET_QUERY)) {
+      next();
+      return;
+    }
+    const url = new URL(req.url, 'http://localhost');
+    const file =
+      fileByPathname.get(url.pathname) ??
+      [...fileByPathname].find(([href]) => href.endsWith(url.pathname))?.[1];
+    if (!file || !url.searchParams.has(STYLESHEET_QUERY)) {
       next();
       return;
     }
```

The match rules do not change: exact match first, then the tail match.

File: packages/unplugin/src/utils/bundledDevCss.ts [hunk 4]
```diff
@@ -0,0 +1,317 @@
...
+      }
+
+      const resolved = await this.resolve(id, importer, { skipSelf: true });
+      if (!resolved || !path.isAbsolute(resolved.id)) return null;
+
+      const file = normalizePath(resolved.id);
+      let holdsMarker = carriesMarker.get(file);
+      if (holdsMarker === undefined) {
+        try {
+          holdsMarker = (await readFile(file, 'utf8')).includes(marker);
```
Comment (Dwlad90) [inline packages/unplugin/src/utils/bundledDevCss.ts]: **O2. Two imports of the same stylesheet read the file twice.**

Rolldown resolves imports in parallel. When two modules import the same stylesheet, both miss `carriesMarker`, both read the full file, and both call `watcher.add`. The result is correct. The work is not needed. Share the read that is in flight for each file.

```diff
+  // Reads in flight, so that the imports of one stylesheet that resolve
+  // together share one read instead of a race for the same answer.
+  const pendingReads = new Map<string, Promise<boolean>>();
 ...
       if (holdsMarker === undefined) {
-        try {
-          holdsMarker = (await readFile(file, 'utf8')).includes(marker);
-        } catch {
-          return null;
-        }
+        let pending = pendingReads.get(file);
+        if (!pending) {
+          pending = readFile(file, 'utf8').then(source => source.includes(marker));
+          pendingReads.set(file, pending);
+        }
+        try {
+          holdsMarker = await pending;
+        } catch {
+          return null;
+        } finally {
+          pendingReads.delete(file);
+        }
         carriesMarker.set(file, holdsMarker);
```

File: packages/unplugin/src/utils/bundledDevCss.ts [hunk 5]
```diff
@@ -0,0 +1,317 @@
...
+
+  // Whether a stylesheet carries the marker, by absolute path. Both answers
+  // are kept and an edit drops the entry, so a file is read once per version.
+  const carriesMarker = new Map<string, boolean>();
+  // Served marker files by absolute path, with the href each is served under,
+  // and the reverse lookup the middleware needs.
+  const hrefByFile = new Map<string, string>();
+  const fileByPathname = new Map<string, string>();
+  // Files the stylesheets `@import`, as Vite's CSS pipeline reports them.
+  const imported = new Set<string>();
```
Comment (Dwlad90) [inline packages/unplugin/src/utils/bundledDevCss.ts]: **O3. `imported` only grows, and the restart list can miss a served file.**

Two related problems with this state.

(a) `imported` is one flat set. The code removes an entry only on `unlink`. When a stylesheet stops to `@import` a file, edits to that file still call `refresh()` for the life of the server.

(b) On restart, `known` comes from `carriesMarker.keys()` and `imported`. An edit deletes the `carriesMarker` entry of the file, and under bundled serve Rolldown does not resolve the file again. So a *served* file can be absent from `known`. The restart makes a new bundle, and that puts the file back, but the bookkeeping here must not depend on that.

Keep the imports for each served stylesheet, and replace that set on each render. Build the watch list from every source of truth.

```diff
-  const imported = new Set<string>();
+  const importsByFile = new Map<string, Set<string>>();
+
+  function isImported(file: string): boolean {
+    for (const imports of importsByFile.values()) {
+      if (imports.has(file)) return true;
+    }
+    return false;
+  }
+
+  function watchedFiles(): string[] {
+    const files = new Set<string>([...hrefByFile.keys(), ...carriesMarker.keys()]);
+    for (const imports of importsByFile.values()) {
+      for (const file of imports) files.add(file);
+    }
+    return [...files];
+  }
 ...
-      for (const dependency of result.deps ?? []) watchImported(normalizePath(dependency));
+      const imports = new Set<string>();
+      for (const dependency of result.deps ?? []) {
+        const imported = normalizePath(dependency);
+        imports.add(imported);
+        server?.watcher.add(imported);
+      }
+      importsByFile.set(file, imports);
 ...
-      const known = [...carriesMarker.keys(), ...imported];
+      const known = watchedFiles();
 ...
-        if (hrefByFile.has(file) || imported.has(file)) refresh();
+        if (hrefByFile.has(file) || isImported(file)) refresh();
```

`forget(file)` then deletes `importsByFile.delete(file)` in place of `imported.delete(file)`. The watcher keeps the old file. `watcher.add` is cheap, and there is no safe way to know that nothing else needs the file.

File: packages/unplugin/src/utils/bundledDevCss.ts [hunk 6]
```diff
@@ -0,0 +1,317 @@
...
+    configure(resolved) {
+      config = resolved;
+      // The client environment's own flag rather than the experimental
+      // option that seeds it: `environments.client.isBundled` can turn
+      // bundled serve on or off independently. Optional all the way down,
+      // since `environments` only exists from Vite 6.
+      enabled =
+        !!marker &&
+        resolved.command === 'serve' &&
+        resolved.environments?.client?.isBundled === true;
```
Comment (Dwlad90) [inline packages/unplugin/src/utils/bundledDevCss.ts:267]: **O4. This new mode has no log output.**

When the styles do not appear, a developer cannot see in the terminal that the plugin is in served-stylesheet mode, or which files it serves. The rest of `index.ts` logs with a `[stylex-unplugin]` prefix.

Render failures are fine as they are. `.catch(next)` reaches Vite's `errorMiddleware`, and that logs `Internal server error: ...` with the stack. I checked this. A second log there is a duplicate.

Add one `info` line when the mode turns on, and one for each served href. Use `config.logger`, so that `logLevel` applies.

```diff
+const LOG_PREFIX = '[stylex-unplugin]';
 ...
       enabled =
         !!marker &&
         resolved.command === 'serve' &&
         resolved.environments?.client?.isBundled === true;
+      if (enabled) {
+        resolved.logger.info(
+          `${LOG_PREFIX} bundled dev server: marker stylesheets are served outside the bundle`
+        );
+      }
 ...
         hrefByFile.set(file, href);
         fileByPathname.set(href, file);
+        config.logger.info(`${LOG_PREFIX} serving ${href} outside the bundle`);
```

File: packages/unplugin/src/utils/bundledDevCss.ts [hunk 7]
```diff
@@ -0,0 +1,317 @@
...
+      const result = await preprocessCSS(stylesheet, file, resolvedConfig);
+      for (const dependency of result.deps ?? []) watchImported(normalizePath(dependency));
+
+      // Vite's own responder: content type, ETag with `304` on a match, and
+      // `HEAD` handling.
+      send(req, res, result.code, 'css', { cacheControl: 'no-store' });
+      // `next` is Connect's error handler here, not a completion callback:
+      // forwarding the rejection to it is the documented middleware contract.
+      // oxlint-disable-next-line promise/no-callback-in-promise
+    })().catch(next);
```
Comment (Dwlad90) [inline packages/unplugin/src/utils/bundledDevCss.ts:250]: **O6. Keep the render for each request. Do not add a cache here.**

Each request reads the file again, runs `transformStyleXCSS` again, and runs `preprocessCSS` again. This looks like a place for a single-flight promise or a memo. It is not.

A refetch starts *after* the plugin records new rules. If two requests share one render that started before those rules, the browser gets old CSS, and it stays old until the next event. This module does not know the rules revision, so it cannot make a safe cache key.

I measured 20 parallel requests for the stylesheet on a small fixture. All of them return `200` in about 500 ms in total. That is acceptable for a dev server. I recommend that you keep this code as it is, and add a short comment that says why there is no cache.

**Review by Dwlad90 — COMMENTED** [review]
**Optimization / scalability / complexity findings for `bundledDevCss.ts`**

The inline comments below cover five points: the request hot path (O1), a duplicate-read race in `resolveId` (O2), the `imported` bookkeeping (O3), logging (O4), and one thing that I recommend you keep as it is (O6). This top-level comment is O5.

**O5. The new module has no tests.**

`bundledDevCss.ts` adds 317 lines and no test file. The PR description lists a manual checklist in a headless Chromium. None of that is in the repository. `guidelines/coding/TESTING.md` asks for high coverage of critical logic and edge cases, and each other file in `src/utils/` has a sibling `__tests__/<name>.test.ts`.

The existing `vite.test.ts` already starts a real Vite server for each fixture, so a bundled-dev suite fits the house style. I wrote one and ran it against a refactor that includes the fixes from these comments. It passes together with the existing Vite and placeholder suites (86 tests). The five cases are:

1. The import of the marker stylesheet becomes the runtime module, and the CSS is not in the bundle.
2. `GET /stylex.css?stylex-css` returns `200`, `text/css`, `no-store`, and the rules at the marker.
3. A non-root base works, and a request by the tail of the href works.
4. A request without the query goes to the next middleware.
5. A render failure goes to the error handler.

Two notes for the test. Middleware mode has no HTTP server, so the test calls `server.middlewares(req, res, next)` with a small mock response. Vite gives no public promise for the first Rolldown build, so the test polls the fallback page (`__vite_is_fallback_page__`) until the bundle is ready. This is the same signal that the browser client uses.

I can push the test file to this branch if you want it.

File: packages/unplugin/README.md [hunk 8]
```diff
@@ -360,8 +360,35 @@ The plugin replaces the marker with the generated StyleX CSS during the build.
...
+> a plain module outside a framework's refresh runtime, reloads the page.
+>
+> Two things stay fixed for the life of the server: which stylesheets carry the
+> marker, since Rolldown resolves an import once, and where the stylesheet sits
+> in the cascade, so adding or removing the marker or the import needs a
+> restart. The marker has to sit in a plain stylesheet imported from
+> JavaScript: a CSS module or a stylesheet used as a build entry is left to
+> Rolldown and only gets the rules known when it is bundled. Assets referenced
+> with `url()` from the marker stylesheet are not rewritten or served in this
+> mode; keep them in `public/` or in a stylesheet without the marker.
```
Comment (Dwlad90) [inline packages/unplugin/README.md]: **Please make this note shorter.**

The note is 24 lines and two paragraphs. It explains how the feature works inside, and it repeats content that is already in `CONTEXT.md` ("Served stylesheet") and in the code comments of `bundledDevCss.ts`. A README reader needs three things: what happens, what updates over HMR, and what the limits are. The rest is implementation detail.

The suggestion below keeps every fact that a user needs, and it removes the explanation of Rolldown, the runtime module, and the cascade position. It also keeps the 80-column limit for markdown. Each limit is now one bullet, so a user can scan the list.

```suggestion
> [!NOTE]
> **Vite's bundled dev server** (`experimental.bundledDev`)
>
> Bundled serve compiles every imported stylesheet into JavaScript. In this
> mode the plugin serves the stylesheet that carries the marker itself, with
> the current rules at the marker. PostCSS, Lightning CSS and `@import` run
> over it as usual. Edits to StyleX modules, to shared variables and themes,
> to the stylesheet, and to the files it `@import`s refetch the stylesheet
> over HMR. Rolldown decides if the edit itself is a hot update or a reload,
> as for any module.
>
> Limits in this mode:
>
> - The marker must be in a plain stylesheet that JavaScript imports. A CSS
>   module or a build entry gets only the rules known at bundle time.
> - The set of marker stylesheets and their position in the cascade are fixed
>   for the life of the server. Restart after you add or remove a marker or an
>   import.
> - `url()` references in the marker stylesheet are not rewritten or served.
>   Keep those assets in `public/` or in a stylesheet without the marker.
```

The three lines added to the paragraph above (363–366, "In development the rules are put into the stylesheet before the pipeline runs ... as described next") can also become one sentence: "In development the rules are put into the stylesheet before the pipeline runs, so there it sees everything." The note below then explains the bundled case.

**Comment by Dwlad90** [comment]
Hi @skovhus 

Thank you for reporting the issue and preparing this PR with the fix!
I also apologize for the AI-generated feedback earlier—I understand it was a bit unclear and confusing.

Once these adjustments are made, I'll be happy to approve and merge this PR.

**Review by Dwlad90 — APPROVED** [review]

**Comment by Dwlad90** [comment]
Thanks @skovhus for the PR!

All changes and fixes are now available in version 0.19.0-rc.2.

Please test it out on your backend to make sure everything works as expected.
